### PR TITLE
[DON'T MERGE]: Testing SAEA hypothesis.

### PR DESCRIPTION
**Pull request by baulig**
**Update 07/30**: Now testing a potential fix.

**Update 07/25**: This is still in use, please don't delete.

### CODE REVIEW

**Comment by baulig** [comment]
@monojenkins pkg

**Comment by garuma** [comment]
I don't see a .pkg entry in the checklist like other builds, I see you requested it but where can I find it in those builds?

Thanks for the PR!

**Comment by baulig** [comment]
@monojenkins build pkg

**Comment by baulig** [comment]
The hypothesis was false.  Trying something else now.

**Comment by baulig** [comment]
@monojenkins build pkg

**Comment by baulig** [comment]
We just got a beautiful test failure on CI:

```
2019-06-25T19:00:34.9535360Z [ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidTimeZoneException: I LIVE ON THE MOON: devdiv.pkgs.visualstudio.com 443
2019-06-25T19:00:34.9536820Z   at System.Environment.get_StackTrace () [0x00000] in <f3865bd7f47847adac3756ce11a84efe>:0 
2019-06-25T19:00:34.9537720Z   at System.Net.Http.ConnectHelper+ConnectEventArgs.OnCompleted (System.Net.Sockets.SocketAsyncEventArgs _) [0x00000] in <859fb3b8513843c68ccb971579b101ff>:0 
2019-06-25T19:00:34.9538420Z   at System.Net.Sockets.SocketAsyncEventArgs.Complete () [0x00000] in <85f34580c9ab4e6499d0c5e1f4caea50>:0 
2019-06-25T19:00:34.9539390Z   at System.Net.Sockets.SocketAsyncEventArgs.FinishOperationAsyncFailure (System.Exception exception, System.Int32 bytesTransferred, System.Net.Sockets.SocketFlags flags) [0x00000] in <85f34580c9ab4e6499d0c5e1f4caea50>:0 
2019-06-25T19:00:34.9540210Z   at System.Net.Sockets.MultipleConnectAsync.AsyncFail (System.Exception e) [0x00000] in <85f34580c9ab4e6499d0c5e1f4caea50>:0 
2019-06-25T19:00:34.9541000Z   at System.Net.Sockets.MultipleConnectAsync.InternalConnectCallback (System.Object sender, System.Net.Sockets.SocketAsyncEventArgs args) [0x00000] in <85f34580c9ab4e6499d0c5e1f4caea50>:0 
2019-06-25T19:00:34.9541660Z   at System.Net.Sockets.SocketAsyncEventArgs.OnCompleted (System.Net.Sockets.SocketAsyncEventArgs e) [0x00000] in <85f34580c9ab4e6499d0c5e1f4caea50>:0 
2019-06-25T19:00:34.9542430Z   at System.Net.Sockets.SocketAsyncEventArgs.Complete () [0x00000] in <85f34580c9ab4e6499d0c5e1f4caea50>:0 
2019-06-25T19:00:34.9542930Z   at System.Net.Sockets.Socket+<>c.<.cctor>b__315_3 (System.IAsyncResult ares) [0x00000] in <85f34580c9ab4e6499d0c5e1f4caea50>:0 
2019-06-25T19:00:34.9543380Z   at System.Net.Sockets.SocketAsyncResult+<>c.<Complete>b__27_0 (System.Object state) [0x00000] in <85f34580c9ab4e6499d0c5e1f4caea50>:0 
2019-06-25T19:00:34.9543870Z   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () [0x00000] in <f3865bd7f47847adac3756ce11a84efe>:0 
2019-06-25T19:00:34.9544400Z   at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00000] in <f3865bd7f47847adac3756ce11a84efe>:0 
2019-06-25T19:00:34.9545110Z   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () [0x00000] in <f3865bd7f47847adac3756ce11a84efe>:0 
2019-06-25T19:00:34.9545620Z   at System.Net.Http.ConnectHelper+ConnectEventArgs.OnCompleted (System.Net.Sockets.SocketAsyncEventArgs _) [0x00031] in <859fb3b8513843c68ccb971579b101ff>:0 
2019-06-25T19:00:34.9546070Z   at System.Net.Sockets.SocketAsyncEventArgs.Complete () [0x00009] in <85f34580c9ab4e6499d0c5e1f4caea50>:0 
2019-06-25T19:00:34.9546580Z   at System.Net.Sockets.SocketAsyncEventArgs.FinishOperationAsyncFailure (System.Exception exception, System.Int32 bytesTransferred, System.Net.Sockets.SocketFlags flags) [0x0001d] in <85f34580c9ab4e6499d0c5e1f4caea50>:0 
2019-06-25T19:00:34.9547090Z   at System.Net.Sockets.MultipleConnectAsync.AsyncFail (System.Exception e) [0x0001a] in <85f34580c9ab4e6499d0c5e1f4caea50>:0 
2019-06-25T19:00:34.9547590Z   at System.Net.Sockets.MultipleConnectAsync.InternalConnectCallback (System.Object sender, System.Net.Sockets.SocketAsyncEventArgs args) [0x000b7] in <85f34580c9ab4e6499d0c5e1f4caea50>:0 
2019-06-25T19:00:34.9548100Z   at System.Net.Sockets.SocketAsyncEventArgs.OnCompleted (System.Net.Sockets.SocketAsyncEventArgs e) [0x0000e] in <85f34580c9ab4e6499d0c5e1f4caea50>:0 
2019-06-25T19:00:34.9548540Z   at System.Net.Sockets.SocketAsyncEventArgs.Complete () [0x00009] in <85f34580c9ab4e6499d0c5e1f4caea50>:0 
2019-06-25T19:00:34.9549310Z   at System.Net.Sockets.Socket+<>c.<.cctor>b__315_3 (System.IAsyncResult ares) [0x00057] in <85f34580c9ab4e6499d0c5e1f4caea50>:0 
2019-06-25T19:00:34.9549780Z   at System.Net.Sockets.SocketAsyncResult+<>c.<Complete>b__27_0 (System.Object state) [0x0000b] in <85f34580c9ab4e6499d0c5e1f4caea50>:0 
2019-06-25T19:00:34.9550240Z   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () [0x00008] in <f3865bd7f47847adac3756ce11a84efe>:0 
2019-06-25T19:00:34.9550920Z   at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00074] in <f3865bd7f47847adac3756ce11a84efe>:0 
2019-06-25T19:00:34.9551350Z   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () [0x00000] in <f3865bd7f47847adac3756ce11a84efe>:0 
2019-06-25T19:00:34.9992020Z ##[error]Error: /Library/Frameworks/Mono.framework/Versions/Current/Commands/msbuild failed with return code: 255
```

**Comment by baulig** [comment]
@monojenkins build pkg

**Comment by baulig** [comment]
There is a test failure:

https://jenkins.mono-project.com/job/test-mono-pull-request-amd64-osx/23099/testReport/junit/(root)/SoftwarePublisherCertificateTest/NonSignedAssembly/

```
SoftwarePublisherCertificateTest.NonSignedAssembly

Failing for the past 1 build (Since Unstable#23099 )
Took 5.3 sec.
Error Message

System.Security.Cryptography.CryptographicException was expected
```

**Comment by baulig** [comment]
@monojenkins build pkg

**Comment by CoffeeFlux** [comment]
Is this still being used?

**Comment by baulig** [comment]
> Is this still being used?

I have essentially given up on the issue more or less - but also don't want to admit defeat that easily.  Let's keep it around for a few more days.

**Comment by baulig** [comment]
It looks like there might be additional data from telemetry which might help us - so let's please keep this open and disregard my last comment.

File: mcs/class/System/System.Net.Sockets/SocketAsyncEventArgs.cs [hunk 1]
```diff
@@ -197,8 +197,14 @@ internal void SetLastOperation (SocketAsyncOperation op)
...
 		}
 
+		int completed;
+
 		internal void Complete ()
 		{
+			var old = Interlocked.CompareExchange(ref completed, 1, 0);
+			if (old != 0)
+				throw new InvalidTimeZoneException ($"I LIVE ON ENCELADUS!");
+
```
Comment (baulig) [inline mcs/class/System/System.Net.Sockets/SocketAsyncEventArgs.cs]: On CI, it was just crashing here!

**Comment by baulig** [comment]
@monojenkins build pkg

**Comment by baulig** [comment]
The bug is finally fixed!

**Pull request closed without merging**